snuba/utils/bucket_timer.py

```diff
@@ -0,0 +1,104 @@
+from __future__ import annotations
+
+from collections import defaultdict, deque
+from datetime import datetime, timedelta
+from typing import Deque, List, Optional
+
+from snuba import environment
+from snuba.utils.metrics.wrapper import MetricsWrapper
+
+WINDOW_SIZE = timedelta(minutes=10)  # minutes
+
+
+class Bucket:
+    def __init__(
+        self,
+        project_id: Optional[int],
+        minute: datetime,
+        processing_time: timedelta = timedelta(seconds=0),
+    ) -> None:
+        self.project_id = project_id  # is None for global bucket
+        self.minute = minute
+        self.processing_time = processing_time
+
+    def add_processing_time_seconds(self, seconds: timedelta) -> None:
+        self.processing_time += seconds
+
+
+class Counter:
+    def __init__(self, consumer_group: str) -> None:
+        self.consumer_group: str = consumer_group
+        self.buckets: Deque[Bucket] = deque()
+
+    def get_time_sum(self, group: List[timedelta]) -> timedelta:
+        return sum([processing_time for processing_time in group], timedelta())
+
+    def floor_minute(self, time: datetime) -> datetime:
+        return time - timedelta(seconds=time.second, microseconds=time.microsecond)
+
+    def ceil_minute(self, time: datetime) -> datetime:
+        if time.second == 0 and time.microsecond == 0:
+            return time
+        return self.floor_minute(time + timedelta(minutes=1))
+
+    def trim_expired_buckets(self, now: datetime) -> None:
+        current_minute = self.floor_minute(now)
+        window = current_minute - WINDOW_SIZE
```

Comment:
nit: rename window with `window_start` or soemthing that tells you that it is the beginning of the window. The window is an interval while you represent a point in time with this variable.